src/node_main_instance.h

```diff
@@ -14,29 +15,77 @@ namespace node {
 // We may be able to create an abstract class to reuse some of the routines.
 class NodeMainInstance {
  public:
-  NodeMainInstance(const NodeMainInstance&) = delete;
-  NodeMainInstance& operator=(const NodeMainInstance&) = delete;
-  NodeMainInstance(NodeMainInstance&&) = delete;
-  NodeMainInstance& operator=(NodeMainInstance&&) = delete;
+  // To create a main instance that does not own the isoalte,
+  // The caller needs to do:
+  //
+  //   Isolate* isolate = Isolate::Allocate();
+  //   platform->RegisterIsolate(isolate, loop);
+  //   isolate->Initialize(...);
+  //   isolate->Enter();
+  //   NodeMainInstance* main_instance =
+  //       NodeMainInstance::Create(isolate, loop, args, exec_args);
+  //
+  // When tearing it down:
+  //
+  //   main_instance->Cleanup();  // While the isolate is entered
+  //   isolate->Exit();
+  //   isolate->Dispose();
+  //   platform->UnregisterIsolate(isolate);
+  //
+  // After calling Dispose() the main_instance is no longer accessible.
+  static NodeMainInstance* Create(v8::Isolate* isolate,
+                                  uv_loop_t* event_loop,
+                                  MultiIsolatePlatform* platform,
+                                  const std::vector<std::string>& args,
+                                  const std::vector<std::string>& exec_args);
+  void Dispose();
 
-  NodeMainInstance(uv_loop_t* event_loop,
-                   const std::vector<std::string>& args,
-                   const std::vector<std::string>& exec_args);
+  // Create a main instance that owns the isolate
+  NodeMainInstance(
+      v8::Isolate::CreateParams* params,
+      uv_loop_t* event_loop,
+      MultiIsolatePlatform* platform,
+      const std::vector<std::string>& args,
+      const std::vector<std::string>& exec_args,
+      const std::vector<size_t>* per_isolate_data_indexes = nullptr);
   ~NodeMainInstance();
 
   // Start running the Node.js instances, return the exit code when finished.
   int Run();
 
- private:
+  IsolateData* isolate_data() { return isolate_data_.get(); }
+
   // TODO(joyeecheung): align this with the CreateEnvironment exposed in node.h
   // and the environment creation routine in workers somehow.
   std::unique_ptr<Environment> CreateMainEnvironment(int* exit_code);
 
+  // If nullptr is returned, the binary is not built with embedded
+  // snapshot.
+  static const std::vector<size_t>* GetIsolateDataIndexes();
+  static v8::StartupData* GetEmbeddedSnapshotBlob();
+
+  static const size_t kNodeContextIndex;
```

Comment:
Just curious, what's the reason you assign it in the .cc file and not here?

Reply:
I guess no? (Or I don't remember any more). I could move the definition here.